src/fiat/drhook/drhook_papi.c

```diff
@@ -0,0 +1,325 @@
+  if (err!=PAPI_OK){
+    printf("DRHOOK:PAPI:PAPI_read: Error reading counters, thread=%ld es=%d %s\n",safe_thread_num(),drhook_papi_event_set[safe_thread_num()],PAPI_strerror(err));
+  }
+#if defined(DEBUG)
+  drhook_papi_print("readAll:",counterArray);
+#endif
+  return err;
+}
+
+void drhook_papi_readall_(long_long * counterArray){
+  drhook_papi_readAll(counterArray);
+}
+
+/* return 1 if papi can be used after the call */
+int drhook_papi_init(int rank){
+  int lib_version;
+  char pmsg[STD_MSG_LEN];
+  int paperr=-1;
+
+  
+  if (drhook_papi_state==drhook_papi_running) return 1;
+  if (drhook_papi_state==drhook_papi_failed) return 0;
+
+  drhook_papi_rank=rank;
+
+#ifdef _OPENMP
+  if (omp_in_parallel()){
+    printf("drhook_papi: Tried to initialise from a parallel region :-(\n");
+    return 0;
+  }
+#endif
+  
+  paperr=PAPI_library_init(PAPI_VER_CURRENT);
+  if (paperr !=  PAPI_VER_CURRENT){
+    snprintf(pmsg,STD_MSG_LEN,"DRHOOK:PAPI:PAPI_library_init: ret=%d ver=%d ",
+	     paperr,PAPI_VER_CURRENT);
+    printf("%s\n",pmsg);
+    if (paperr > 0) {
+      snprintf(pmsg,STD_MSG_LEN,"DRHOOK:PAPI: library version mismatch!\n");
+      printf("%s\n",pmsg);
+      return 0;
+    }
+    if (paperr == PAPI_EINVAL){
+      snprintf(pmsg,STD_MSG_LEN,"DRHOOK:PAPI: PAPI_EINVAL\n");
+      printf("%s\n",pmsg);
+      return 0;
+    }
+    if (paperr == PAPI_ENOMEM){
+      snprintf(pmsg,STD_MSG_LEN,"DRHOOK:PAPI: PAPI_ENOMEM\n");
+      printf("%s\n",pmsg);
+      return 0;
+    }
+    if (paperr == PAPI_ESBSTR){
+      snprintf(pmsg,STD_MSG_LEN,"DRHOOK:PAPI: PAPI_ESBSTR\n");
+      printf("%s\n",pmsg);
+      return 0;
+    }
+    if (paperr == PAPI_ESYS){
+      snprintf(pmsg,STD_MSG_LEN,"DRHOOK:PAPI: PAPI_ESYS\n");
+      printf("%s\n",pmsg);
+      return 0;
+    }
+    else {
+      snprintf(pmsg,STD_MSG_LEN,"DRHOOK:PAPI: Unknown error code\n");
+      printf("%s\n",pmsg);
+      return 0;
+    }
+  }
+    
+  lib_version = PAPI_get_opt( PAPI_LIB_VERSION, NULL );
+
+  int nthreads=1;
+#ifdef _OPENMP
+  nthreads=omp_get_max_threads();
+#endif
+
+  paperr=PAPI_thread_init(safe_thread_num);
+  
+  if( paperr != PAPI_OK ){
+    snprintf(pmsg,STD_MSG_LEN,"DRHOOK:PAPI: thread init failed (%s)",PAPI_strerror(paperr));
+    printf("%s\n",pmsg);
+    return 0;
+  }
+
+    snprintf(pmsg,STD_MSG_LEN,"DRHOOK:PAPI: Version %d.%d.%d initialised with %d threads",
+	     PAPI_VERSION_MAJOR( lib_version ),
+	     PAPI_VERSION_MINOR( lib_version ),
+	     PAPI_VERSION_REVISION( lib_version ),
+	     nthreads );
+  
+    if (drhook_papi_rank==0) printf("%s\n",pmsg);
+
+  drhook_papi_event_set=malloc(nthreads*sizeof(int));
+
+  int prof_papi_numcntrs;
+  bool failed=false;  
+#pragma omp parallel default(none) shared(drhook_papi_rank,drhook_max_counter_name,hookCounters,drhook_papi_event_set,nthreads,prof_papi_numcntrs,failed)
+  {
+    int paperr2,counter;
+    int * events;
+    int number;
+    int j=1;
+#ifdef _OPENMP
+    j=safe_thread_num();
```

Comment:
`safe_thread_num()` normally returns 0 when `_OPENMP` is not defined.
So I thought it seems redundant to have `j=safe_thread_num()` wrapped with `#ifdef _OPENMP`.
However then the default value `j=1` would become `j=0`.
That would bypass the further `j==k` condition when `nthreads=1` as is the case without `_OPENMP` defined.
Is that the intention, why `j=1`?